Geotrellis demo framework

tnation14 opened the pull request:
# Overview

Bring this project into alignment with the GeoTrellis Demo Framework used for the Chattanooga, Transit, And PointCloud demos. This allows us to automatically deploy the RasterVision service to our GeoTrellis ECS cluster after merges to master. I also made some naming updates in the development environment, and configured nginx to redirect HTTP requests to HTTPS.

# Notes
- Nginx has a healthcheck location block configured, but the backend doesn't have a `health-check` endpoint. In order to keep services from being recycled, I just used the default `/` path  & `200`response healthcheck from the terraform module. 

- I moved the `node_modules` cache to `${PWD}/rastervision-node-modules` because the Travis CI user doesn't have write permissions to `/var/cache/`. 

- I enabled the nginx HTTP --> HTTPS stuff for the sake of consistency, but I don't think this project is using the http port at all. In development, requests go directly to the HTTPS port, and in Staging, CloudFront and the ALB only listen on HTTPS. The HTTP redirect can probably be ripped out.

## Notable changes
- Unpin patch version for `awscli`
- Changed shared folder to NFS for faster build times
- Add HTTP --> HTTPS redirect to nginx config
- Use `terraform-aws-ecs-web-service` module to create ECS resources
- Add security group to allow ALB to communicate with the outside world over HTTPS, and the Container Instance
- Add CDN to cache tiles
- Add travis-ci deployment configuration
- Fix various CI build problems (static files weren't captured in the `nginx`  container, fix broken webpack Favicon configuration, stale references to PointCloud demo)

# Testing
## Local tests
- Bring up a new VM with `scripts/setup`, then run `scripts/server` inside of the VM.
- Test `nginx` by making sure you can access http://localhost:9100. 
- Test the webpack proxy config by accessing http://localhost:8284. All tiles should render.

## CI tests
- See Travis Build output (TBD)

## Terraform tests
- see https://potsdam.geotrellis.io and https://potsdam-origin.geotrellis.io

Closes #11

hectcastro commented on Vagrantfile:
```diff
@@ -3,17 +3,25 @@
 
   config.vm.synced_folder "~/.aws", "/home/vagrant/.aws"
   config.vm.synced_folder "~/.ivy2", "/home/vagrant/.ivy2"
+  config.vm.synced_folder "./", "/vagrant", type: "nfs", mount_options: MOUNT_OPTIONS
```
This reminds me of our discussion earlier today where `.` was being mounted multiple times in different areas via different mechanisms. Here we are likely skirting that issue because we are overriding the default.

hectcastro commented on Vagrantfile:
```diff
@@ -29,7 +37,7 @@ Vagrant.configure(2) do |config|
 
   config.vm.provision "ansible" do |ansible|
-    ansible.playbook = "deployment/ansible/pc-demo.yml"
+    ansible.playbook = "deployment/ansible/rastervision.yml"
```
I don't know enough to know whether this rename is accurate or not. Are you sure it makes sense to name things here after Raster Vision generically?

tnation14 replied:
I thought so because all of the pre-existing stuff referred to the project as Raster Vision. Maybe @lossyrob or @lewfish can give a more definitive answer.

lewfish replied:
This app serves data generated by raster-vision, but is not raster-visions itself, so perhaps isprs-potsdam-viz is a better name. I think raster-vision should just be used to refer to the raster-vision repo.

hectcastro commented on deployment/terraform/app.tf:
```diff
@@ -1,121 +1,50 @@
   vars {
-    nginx_image_url      = "279682201306.dkr.ecr.us-east-1.amazonaws.com/rastervision-nginx:${var.image_version}"
-    api_server_image_url = "279682201306.dkr.ecr.us-east-1.amazonaws.com/rastervision-api-server:${var.image_version}"
+    nginx_image_url      = "896538046175.dkr.ecr.us-east-1.amazonaws.com/rastervision-nginx:${var.image_version}"
```
The other demos have this as a `aws_account_id ` variable. We should likely do the same here.

hectcastro commented on deployment/terraform/app.tf:
```diff
@@ -1,121 +1,50 @@
+  deployment_min_healthy_percent = "${var.raster_vision_ecs_deployment_min_percent}"
+  deployment_max_percent         = "${var.raster_vision_ecs_deployment_max_percent}"
+  container_name                 = "nginx"
+  container_port                 = "443"
```
An explicit `health_check_path` should probably be set here to avoid an issue where defaults don't work in the future.

hectcastro commented on deployment/terraform/cdn.tf:
```diff
@@ -0,0 +1,62 @@
+      headers      = ["*"]
+
+      cookies {
+        forward = "none"
```
In the tree prioritization demo we are passing all of the cookies through. I don't think this application makes use of cookies, but we should pass them through anyway because if we do at some point in the future, this may be hard to catch.

hectcastro commented on deployment/terraform/firewall.tf:
```diff
@@ -0,0 +1,55 @@
+  security_group_id = "${module.raster_vision_ecs_service.lb_security_group_id}"
+}
+
+resource "aws_security_group_rule" "alb_http_ingress" {
```
Is this necessary if we're using CloudFront to handle HTTP -> HTTPS?

tnation14 replied:
Ah, it isn't.

hectcastro commented on deployment/terraform/firewall.tf:
```diff
@@ -0,0 +1,55 @@
+  security_group_id = "${module.raster_vision_ecs_service.lb_security_group_id}"
+}
+
+resource "aws_security_group_rule" "alb_container_instance_all_egress" {
```
I think that it would be good to reference RV somewhere in this resource name.

hectcastro commented on deployment/terraform/firewall.tf:
```diff
@@ -0,0 +1,55 @@
+#
+# Container instance security group resources
+#
+resource "aws_security_group_rule" "container_instance_alb_all_ingress" {
```
Same here regarding the RV name.

hectcastro commented on deployment/terraform/firewall.tf:
```diff
@@ -0,0 +1,55 @@
+  source_security_group_id = "${module.raster_vision_ecs_service.lb_security_group_id}"
+}
+
+resource "aws_security_group_rule" "container_instance_alb_all_egress" {
```
Same here.

hectcastro commented on deployment/terraform/variables.tf:
```diff
@@ -6,71 +6,59 @@ variable "project_id" {
+  default = "arn:aws:iam::aws:policy/AmazonS3FullAccess"
+}
 
 variable server_app_alb_ingress_cidr_block {
```
This doesn't seem to be used any more.

hectcastro commented on deployment/terraform/variables.tf:
```diff
@@ -6,71 +6,59 @@ variable "project_id" {
   default = "RV"
```
This doesn't seem to be used any more.

hectcastro commented on src/nginx/etc/nginx/conf.d/default.conf:
```diff
@@ -2,20 +2,15 @@ upstream api-server-upstream {
-# with webpack dev server's proxying.
-# Failed with error of ERPROTO; looks like an SSL problem
-# when referring to nginx by it's docker name.
+server {
```
It doesn't seem like we have anything bound to this port anymore.

hectcastro commented on Vagrantfile:
```diff
@@ -3,17 +3,25 @@
   config.vm.box = "ubuntu/trusty64"
 
   config.vm.synced_folder "~/.aws", "/home/vagrant/.aws"
   config.vm.synced_folder "~/.ivy2", "/home/vagrant/.ivy2"
```
Omitting this share might be good too. Tough tradeoff because we lose access to the host cache, but also avoid potentially reading and writing a bunch over the shared file system.

hectcastro commented on docker-compose.yml:
```diff
@@ -18,12 +18,15 @@ services:
         # Ensure node_modules cache doesn't clash with other jobs on CI.
-        - /var/cache/pc-demo-node-modules:/usr/src/node_modules
-    command: npm run dev-server
+        - ${PWD}/potsdam-demo-node-modules:/usr/src/node_modules
```
I think that this volume mount is overlapping with the first (`./src/app-frontend:/usr/src`). After running `update`, I'm seeing multiple `node_modules` directories:

```bash
$ git clean -ndx --exclude .vagrant | grep node
Would remove potsdam-demo-node-modules/
Would remove src/app-frontend/node_modules/
```

hectcastro commented on docker-compose.yml:
```diff
@@ -33,10 +36,10 @@ services:
         - ./data/catalog-hadoop:/opt/catalog-hadoop/
         - ./data/catalog:/opt/catalog/
         - ./src/app-backend/.sbt:/root/.sbt
```
I think that this is another scenario where we're volume mounting a parent and then a child directory inside of it. It appears to be leading to:

```
api-server_1    | Caused by: java.lang.ClassNotFoundException: sbt.Tracked$
api-server_1    |       at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
api-server_1    |       at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
api-server_1    |       at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
api-server_1    |       at sbt.Classpaths$$anonfun$dependencyPositionsTask$1.apply(Defaults.scala:1561)
api-server_1    |       at sbt.Classpaths$$anonfun$dependencyPositionsTask$1.apply(Defaults.scala:1536)
api-server_1    |       at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
api-server_1    |       at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
api-server_1    |       at sbt.std.Transform$$anon$4.work(System.scala:63)
api-server_1    |       at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
api-server_1    |       at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
api-server_1    |       at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
api-server_1    |       at sbt.Execute.work(Execute.scala:237)
api-server_1    |       at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
api-server_1    |       at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
api-server_1    |       at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
api-server_1    |       at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
api-server_1    |       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
api-server_1    |       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
api-server_1    |       at java.util.concurrent.FutureTask.run(FutureTask.java:266)
api-server_1    |       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
api-server_1    |       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
api-server_1    |       at java.lang.Thread.run(Thread.java:748)
api-server_1    | [error] (*:dependencyPositions) java.lang.NoClassDefFoundError: sbt/Tracked$
api-server_1    | Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? vagrant_api-server_1 exited with code 1
```
If I comment this line, the service starts. Before commenting it out, I made sure that I cleared all cache and target directories to ensure a clean slate.

hectcastro commented on Vagrantfile:
```diff
@@ -3,17 +3,25 @@
 
 Vagrant.require_version ">= 1.8"
 
+MOUNT_OPTIONS = if Vagrant::Util::Platform.linux? then
```
I ended up replacing NFS with `rsync` to get around issues with classes not found during the backend compilation phases. Still unsure of what exactly is going on, but the following diff got me to a working state:

```diff
diff --git a/Vagrantfile b/Vagrantfile
index fcd404d..6723805 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -3,17 +3,14 @@
 
 Vagrant.require_version ">= 1.8"
 
-MOUNT_OPTIONS = if Vagrant::Util::Platform.linux? then
-                  ['rw', 'vers=4', 'tcp', 'nolock']
-                else
-                  ['vers=3', 'udp']
-                end
 Vagrant.configure(2) do |config|
   config.vm.box = "ubuntu/trusty64"
 
   config.vm.synced_folder "~/.aws", "/home/vagrant/.aws"
-  config.vm.synced_folder "~/.ivy2", "/home/vagrant/.ivy2"
-  config.vm.synced_folder "./", "/vagrant", type: "nfs", mount_options: MOUNT_OPTIONS
+  config.vm.synced_folder ".", "/vagrant", type: "rsync",
+    rsync__exclude: [".git/", "app-backend/**/target/",
+                     "deployment/ansible/roles/azavea*/"],
+    rsync__args: ["--verbose", "--archive", "-z"]
 
   config.vm.provider :virtualbox do |vb|
     vb.memory = 8192
diff --git a/docker-compose.yml b/docker-compose.yml
index 2c522e7..d6bbc3e 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -23,8 +23,6 @@ services:
     volumes:
         - ./src/app-frontend:/usr/src
         - ./src/nginx/srv/dist:/usr/src/dist
-        # Ensure node_modules cache doesn't clash with other jobs on CI.
-        - ${PWD}/potsdam-demo-node-modules:/usr/src/node_modules
     entrypoint: npm
     command: run dev-server
   api-server:
@@ -35,7 +33,6 @@ services:
         - ./src/app-backend/:/opt/app-backend/
         - ./data/catalog-hadoop:/opt/catalog-hadoop/
         - ./data/catalog:/opt/catalog/
-        - ./src/app-backend/.sbt:/root/.sbt
         - $HOME/.ivy2/:/root/.ivy2
         - $HOME/.aws:/root/.aws:ro
     environment:
```

hectcastro reviewed (commented):
I created https://github.com/azavea/operations/issues/146 so that we can tear down the previous version of this stack inside the R&D account.

hectcastro commented on Vagrantfile:
```diff
@@ -7,13 +7,17 @@ Vagrant.configure(2) do |config|
-  config.vm.synced_folder "~/.ivy2", "/home/vagrant/.ivy2"
-
+  config.vm.synced_folder "./", "/vagrant", type: "rsync",
+    rsync__exclude: ["app-backend/**/target/",
```
I think that we need to make sure that `.git` is ignored as well.

tnation14 replied:
Ah, I left it in there so `scripts/cibuild` and `scripts/test` didn't break but I guess people can just run those scripts with `GIT_COMMIT=latest`.

hectcastro commented on Vagrantfile:
```diff
@@ -7,13 +7,17 @@ Vagrant.configure(2) do |config|
     vb.cpus = 4
   end
 
+  config.vm.network "private_network", ip: ENV.fetch("RV_PRIVATE_IP", "10.10.10.10")
```
Is this still necessary since we removed NFS support?

hectcastro commented on deployment/terraform/storage.tf:
```diff
@@ -1,18 +1,12 @@
-#   }
-# }
 
+resource "aws_s3_bucket" "catalogs" {
```
Is something supposed to be in this bucket?

tnation14 replied:
Not currently, no. The app is configured to read catalogs that live at `s3://otid-data/viz/catalogs`, but I thought that there might've been some plans to move that data since there was a (commented-out) TF config for this bucket.

hectcastro replied:
@lewfish, does it make sense to copy just the contents of this subdirectory over to a bucket within the GeoTrellis AWS account? The goal is to try and keep things this service depends on within the same account, despite the fact that the GeoTrellis AWS account can read/write to the R&D account.

lewfish replied:
That makes sense to me, but I had very little to do with this repo, so I think we should see what @lossyrob thinks.

lossyrob replied:
Yep, there's no issue with copying this data over to the GeoTrellis account.

hectcastro commented on deployment/terraform/cdn.tf:
```diff
@@ -0,0 +1,62 @@
+
+  enabled          = true
+  http_version     = "http2"
+  comment          = "Geotrellis Potsdam Demo (${var.environment})"
```
s/Geotrellis/GeoTrellis/

Also, it might make sense to omit that altogether because we're already within that account.

hectcastro commented on .gitignore:
```diff
@@ -30,6 +30,7 @@ deployment/terraform/.terraform
 
 # Node and Webpack
 node_modules/
+potsdam-demo-node-modules/
```
Unclear, but doesn't seem like this is necessary any longer.

hectcastro reviewed (approved):
👍 

I think that it would be good to hold off on merging under the home of the catalogs used by this demo is sorted out.

lewfish commented:
Sorry I didn't respond til now. For some reason the notifications for this got routed to a folder I never look at. I think it would be better for Rob to review this than me since he's much more familiar with it.

lossyrob commented:
+1 on moving the catalog data if that simplifies things.

tnation14 committed:
Update development environment configuration

- Add entrypoint for app-frontend container
- Use rsync folder for faster builds
- Loosen ansible version constraints
- Use latest AWS CLI
- Rename references to rastervision to potsdam

tnation14 committed:
Update terraform deployment to use Geotrellis Demo Framework

- Use terraform-aws-ecs-service module to create new ECS task
- ECS task logs to CloudWatch
- Use terraform remote state to access core infrastructure resources
- Remove unnecessary logs bucket
- Add security rules to allow HTTP(s) to the loadbalancers, and all communication between ALB and container instances
- Add CloudFront distribution to cache tiles

tnation14 committed:
Add configuration for Travis CI deployments

tnation14 committed:
Update GeoTrellis catalog location

tnation14 committed:
Webpack should reverse-proxy requests to nginx...

...and output files in the nginx folder
